README.md
# uws-reverse-proxy

This project is an **easy-to-use 0-dependency**\* reverse proxy based on [uWebSockets.js](https://github.com/uNetworking/uWebSockets.js/). It enables use of `uWebSockets.js` and any `http`
server (as [express](https://www.npmjs.com/package/express), [NestJS](https://nestjs.com/), [Fastify](https://www.fastify.io/)) on **the same port**.

Tested with:

- uWebSockets.js v20.10.0
- NodeJS v18.0.0

\* _This package doesn't even depend on `uWebSockets.js` not to force any version usage, see in examples below._
_That said, keep in mind that only the last version of uWebSockets.js has been tested._

## How does it work?

It basically uses `uWebSockets.js` as a reverse proxy for all non-websocket traffic and forward it to any
`http` server.

Then it will send the response back to the client without modification.

## Why? For what use case?

Don't use this package if you don't **really** need it.

You pretty much always want your NodeJS app available on only one port (**443**) for it to be accessible in restrictive **NATs**.

This project is aimed to provide a solution in **restricted server environments** such
as some cloud platforms like Heroku, where you can't set up a proxy like **Nginx** or **Apache** in front of your NodeJS application,
and where you often have only one public port available.

If you're not in a similar use case, just set up an **Nginx** or **Apache** proxy that will filter the requests to redirect them on
the good **private** port, they will be more efficient and flexible.

Conversely, if your server has to work in **restricted server environment** or if for whatever reason you want `uWebSockets.js`
to handle all requests without standing behind a proxy, then this package is made for you :)

In the case of **express** you _could_ use a package like [http-proxy-middleware](https://www.npmjs.com/package/http-proxy-middleware)
to do the opposite (using `express` as a reverse proxy to forward requests to `uWebSockets.js`), but this doesn't seem to work at the time
I'm writing this, and **it defeats the main advantage of `uWebSockets.js`**: its astonishing performances.

## Important note about SSL

If `uWebSockets.js` and your `http` servers are on the same machine, you should not create an HTTPS server,
as TLS will add a useless overhead in a scenario like this:

```mermaid
sequenceDiagram
	participant C as Client
	box Public
	participant UWS as uWebSockets.js Reverse Proxy
	end
	box Private
	participant N as Node:http
	end
	C->>UWS: HTTPS request
	activate UWS
	UWS->>N: HTTP request
	activate N
	N->>UWS: HTTP Response
	deactivate N
	UWS->>C: HTTPS Response
	deactivate UWS
```

## Installation

With npm:

```bash
npm install uws-reverse-proxy
```

With yarn:

```bash
yarn add uws-reverse-proxy
```

## Usage

This section describes some usage scenario. You can even see them in action in the
[examples repository](https://github.com/jordan-breton/uws-reverse-proxy-examples).

To see all available options, check the [code API documentation](https://jordan-breton.github.io/uws-reverse-proxy/).

### Basic

Simplest use-case: let the reverse proxy create everything on its own, configure your servers afterward.

```js
const uWebSockets = require('uWebSockets.js');
const http = require('http');

const {
	UWSProxy,
	createUWSConfig
} = require('uws-reverse-proxy');

const port = process.env.PORT || 80;

const proxy = new UWSProxy(
	createUWSConfig(
		uWebSockets,
		{ port }
	)
);

const httpServer = http.createServer((req, res) => {
	console.log('Incoming request: ', req);

	res.writeHead(200);
	res.end('Hello world :)');
});
httpServer.listen(
	proxy.http.port,
	proxy.http.host,
	() => console.log(`HTTP Server listening at ${proxy.http.protocol}://${proxy.http.host}:${proxy.http.port}`)
);

proxy.start();

// Configuring the uWebSockets.js app created by the proxy

// Setting up websockets
proxy.uws.server.ws({
	upgrade : () => {
		//...
	}
	// ...
});

// Listening
proxy.uws.server.listen('0.0.0.0', port, listening => {
	if(listening){
		console.log(`uWebSockets.js listening on port 0.0.0.0:${port}`);
	}else{
		console.error(`Unable to listen on port 0.0.0.0:${port}!`);
	}
});
```

Second use case: providing already created and listening `uWebSocket.js` server.

```js
const http = require('http');
const { App } = require('uWebSockets.js');

const port = process.env.PORT || 80;
const httpPort = process.env.HTTP_PORT || 35794;
const httpHost = process.env.HTTP_HOST || '127.0.0.1';

// Let's imagine that all the setup happen there, including listening
const httpServer =  http.createServer(/*...*/);
httpServer.listen(httpPort, httpHost, () => console.log(`HTTP server listening at http://${httpHost}:${httpPort}`));

const uwsServer = App();
uwsServer.ws({/* ... */});
uwsServer.listen('0.0.0.0', port, () => { /* ... */ });

const {
	UWSProxy,
	createUWSConfig,
	createHTTPConfig 
} = require('uws-reverse-proxy');

const proxy = new UWSProxy(
	createUWSConfig(
		uwsServer,
		{ port } // Must be specified to avoid a warning
	),
	createHTTPConfig(
		{
			port: httpPort, 
			host: httpHost
		}
	)
);
proxy.start();
```

The warning is there as a reminder especially if you use different ports than the default ones.

If you don't want (or can't) provide the `port`, you can silence the warning using the `quiet` option:

```js
const proxy = new UWSProxy(
	createUWSConfig(
		uwsServer,
		{ quiet: true }
	)
);
proxy.start();
```

### With SSL

Given the above examples, just change the options passed to `createUWSConfig`.

#### Letting the proxy creating the `uWebsockets.js SSLApp`:

```js
const uWebSocket = require('uWebSockets.js');
const { createUWSConfig } = require('uws-reverse-proxy');

createUWSConfig(
	uWebSocket,
	{
		config: {
			key_file_name: 'misc/key.pem',
			cert_file_name: 'misc/cert.pem'
		}
	}
)
```

#### Providing a `uWebsockets.js SSLApp`:

```js
const { SSLApp } = require('uWebSockets.js');
const { createUWSConfig } = require('uws-reverse-proxy');

const uwsServer = SSLApp({
	key_file_name: 'misc/key.pem',
	cert_file_name: 'misc/cert.pem'
});

createUWSConfig(
	uwsServer,
	{ quiet: true }
)
```

### Integration examples with Express, Koa, Fastify, Nestjs, Node:http

Please visit the [examples repository](https://github.com/jordan-breton/uws-reverse-proxy-examples).

It even contains a standalone mode to help you test `uws-reverse-proxy` with your own app without
changing a line of code for most projects.

**Keep in mind that it's... an HTTP reverse proxy. It can forward request to any HTTP server.**

## A word about performances

Putting a proxy in front of your application always induces a performance penalty. This performance penalty heavily depends
on the way you integrate the proxy to your application.

Just keep in mind that contrary to using a proxy like Apache or Nginx, uws-reverse-proxy will only impact the HTTP part
of your application. Since it is uWebSockets.js that handle all the trafic, your websockets are directly exposed to the Internet.

This proxy is especially useful for websockets heavy applications in the uses cases you can find in the
[Why? For what use cases?](#why-for-what-use-cases) section.

### Performance penalty in function of the integration mode

They are two main ways to integrate the proxy to your application:

- **Standalone mode**: the most efficient. The proxy and your application are running in two different processes.
- **Integrated mode**: the least efficient. The proxy and your application are running in the same process.

While you can expect almost no performance penalty in standalone mode, in integrated mode, the performance penalty
is about **30% to 45%**.

In standalone mode, the performance penalty is comprised between **3% to 10%** as long as your HTTP server maximum capacity do not exceed this proxy maximum capacity.

As a matter of fact, in my test environment, the proxy is able to handle up to **45 000 req/s** while an **Express** server is only
able to handle **9 500 req/s**. So, in this case, the proxy is able to handle more requests than the Express server: you'll see almost
no performance penalty.

But if you test the proxy with an **uWebSockets.js** backend server (which is stupid, but it's to illustrate the point),
while the proxy is able to handle up to **45 000 req/s**, the **uWebSockets.js** server is able to handle more or less 250 000 req/s.
In this case, you end up with a **500%** performance penalty.

I advise you to test the proxy to be sure that it fits your needs and that it's not a bottleneck in your specific case.

That being said, you'll find below a table with the performance penalty I measured in my test environment for the different
frameworks I tested. You'll see an **average** in both **standalone** and **integrated** mode as well as **no proxy** at all.

### Performance comparison

Environment:

| Component | Specifications                                    |
| --------- | ------------------------------------------------- |
| OS        | Ubuntu 22.04.2 LTS 64 bits                        |
| Node      | v18.15.0                                          |
| RAM       | 16 Gio                                            |
| Processor | 11th Gen Intel® Core™ i7-11800H @ 2.30GHz × 16 |

No DNS resolution, all servers will listen on the loopback address.

All results are the average of 5 runs of 10 seconds each by using
[autocannon](https://github.com/mcollina/autocannon) with the following command:

```bash
autocannon localhost:7777
# Or, when testing without proxy
autocannon localhost:35794
```

Benchmarks are made with the [uws-reverse-proxy-examples repository](https://github.com/jordan-breton/uws-reverse-proxy-examples)

- In **standalone mode**, the proxy is started with `PORT=7777 yarn standalone:reverse-proxy`, then the target HTTP
  server is started with `yarn standalone:[framework]`.
- In **integrated mode**, we start the server with the command `PORT=7777 yarn [framework]`.

#### Results

| Framework                 |    No proxy | Standalone mode | Integrated mode |
| ------------------------- | ----------: | --------------: | --------------: |
| None (native http module) | 18024 req/s |     17000 req/s |      9803 req/s |
| Koa                       | 13079 req/s |     12400 req/s |      8492 req/s |
| Fastify                   | 12372 req/s |     11284 req/s |      7817 req/s |
| Nestjs                    |  9786 req/s |      8813 req/s |      6153 req/s |
| Express                   |  9362 req/s |      9102 req/s |      6375 req/s |

For readability, I've rounded the results to the nearest integer. A better view of the performance penalty can be
found in the table below based on the results above.

| Framework                 | Standalone mode | Integrated mode |
| ------------------------- | --------------: | --------------: |
| None (native http module) |           5.68% |          45.61% |
| Koa                       |           5.20% |          35.10% |
| Fastify                   |           8.80% |          36.81% |
| Nestjs                    |           9.94% |          37.12% |
| Express                   |           2.78% |          31.91% |

## TODO

- [X]  PoC (> v1.0.0)
- [X]  Refactoring + Clean & stable implementation (>= v2.0.2)
  - [X]  Flexible configuration
  - [X]  Config validation
  - [X]  Config warnings
  - [X]  Configurable backpressure threshold
  - [X]  Code comments & JSDOC
- [X]  Generic reverse proxy for any HTTP/HTTPS server
- [X]  Documentation
  - [X]  Code API
  - [X]  README
  - [X]  A demo repository.
- [X]  Publish on NPM
- [X]  Better error management
  - [X]  Allow answering stream errors that are happening before the client response is written
    with proper HTTP formatted response instead of shutting down the connection like a savage.
  - [X]  Allow for errors customization
- [X]  When a content-length header is present in the response, using uWebSockets.js `tryEnd` instead
  of `write`to avoid adding `Transfer-Encoding: chunked` to every request.
- [ ]  Edge cases handling regarding proxying (HTTP 100 continue, headers cleanup)
- [X]  Testing a `node:http` client to [`undici`](https://github.com/nodejs/undici) client migration to see performances impact.
- [X]  Custom HTTP client implementation to avoid using `node:http` client.
- [ ]  More flexibility in requests routing through proxy (add more control options, like a pre-handler to allow or not
  forwarding based on custom logic.)
- [ ]  Test uWebSockets.js version agnosticity for all uWebSockets.js versions.
  - [ ]  Support backward compatibility
- [ ]  Automated tests included in CI/CD
  - [X]  HTTP/1.1 parser tests
- [ ]  Debugging mode with console logging


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
### Added
- Eslint configuration + compliance

## 3.2.1 - 2023-04-22
### Fixed
- GitHub workflow regarding auto documentation generation not added to the commit.

## 3.2.0 - 2023-04-22
### Added
- A more performant HTTP client supporting HTTP pipelining, Keep-Alive and connection pooling to replace `node:http` (see #2)
- Error handling (you can handle errors and customize the error response) (see #2)
- Started test coverage.
    - Unit Tests for the HTTP Response parser

### Fixed
- Huge performance issue (see #2) with a custom HTTP Client & HTTP Response parser
    - There is still room for improvement, but it's now more than acceptable since the proxy handle
      more connection than most NodeJS HTTP servers.

## 3.1.1 - 2023-03-23
### Changed
- Changed code licensing from AGPL 3.0 and later to LGPL 3.0 or later. This change is retroactive.

## 3.1.0 - 2023-03-22
### Added
- Code API now generated with typedoc
- Added typedoc generation to GitHub action
- Added error response customisation

### Removed
- Integration examples with express, fastify, koa and nestjs. It have been replaced by a link to
  the examples repository.

### Fixed
- Error handling. No more connection shutdown without trying to send a proper error response.
- Various comments to fit the new documentation generation tool.

## 3.0.3 - 2023-03-17
### Fixed
- Install section in documentation (now use the npm package instead of GitHub repository)
- Outdated examples in README
- README spell checked
- Code-api and comments spell checked

## 3.0.2 - 2023-03-17
### Added
- Changelog


src/UWSProxy.js
// region Imports

const HTTPClient = require('./http/Client');
const { isPromise } = require('util').types;

const {
	decodeRequest,
	writeHeaders
} = require('./utils/uwsHelpers');

// endregion

// region Private declarations

/**
 * List of keys in uWebSocket config object that indicates that our server is using SSL encryption.
 * @type {string[]}
 * @private
 */
const UWS_SSL_KEYS = [
	'key_file_name',
	'cert_file_name'
];

// endregion
// region JSDOC typedefs

/**
 *
 * HTTP configuration options for UWSProxy's constructor.
 * @see {UWSProxy.createHTTPConfig}
 * @typedef UWSProxyHTTPConfigOpts
 * @property {'http'|'https'} [protocol="http"] Default: `'http'` - Server protocol
 * @property {int}            [port=35974] Default: `35974` - Private port the HTTP server must listen to
 * @property {string}         [host="127.0.0.1"] Default: `'127.0.0.1'` - HTTP host. Default is the loop-back address.
 * @property {boolean}        [quiet=false] Default: `false` - Disable configuration warning printing
 * @property {UWSClientOpts}  [client] Default: `undefined` - HTTP client configuration
 */

/**
 * UWSProxy actual configuration for uWebSockets.js
 * @typedef UWSProxyHTTPConfig
 * @property {UWSProxyHTTPConfigOpts} config Raw configuration passed to UWSProxy.createHTTPConfig
 * @property {'http'|'https'} protocol HTTP protocol
 * @property {int}            port     Listening port
 * @property {string}         host     HTTP host
 * @property {UWSClientOpts}  client
 */

/**
 * uWebSockets.js
 * @see [uWebSockets.js](https://github.com/uNetworking/uWebSockets.js/)
 * @typedef {import("uWebSockets.js")} UWS
 */

/**
 * uWebSockets.js TemplatedApp
 * @øee [TemplatedApp Documentation](https://unetworking.github.io/uWebSockets.js/generated/interfaces/TemplatedApp.html)
 * @typedef {import("uWebSockets.js").TemplatedApp} UWSTemplatedApp
 */

/**
 * uWebSockets.js AppOptions
 * @see [AppOptions Documentation](https://unetworking.github.io/uWebSockets.js/generated/interfaces/AppOptions.html)
 * @typedef {import("uWebSockets.js").AppOptions} UWSAppOptions
 */

/**
 * uWebSockets.js HTTPRequest
 * @see [HttpRequest Documentation](https://unetworking.github.io/uWebSockets.js/generated/interfaces/HttpRequest.html)
 * @typedef {import("uWebSockets.js").HttpRequest} UWSRequest
 */

/**
 * uWebSockets.js Recognized string
 * @see [RecognizedString Documentation](https://unetworking.github.io/uWebSockets.js/generated/types/RecognizedString.html)
 * @typedef {import("uWebSockets.js").RecognizedString} UWSRecognizedString
 */

/**
 * uWebSockets.js HttpResponse
 * @see [HttpResponse Documentation](https://unetworking.github.io/uWebSockets.js/generated/interfaces/HttpResponse.html)
 * @typedef {import("uWebSockets.js").HttpResponse} UWSResponse
 */

/**
 * HTTPClient configuration options.
 * @typedef {import("./http/Client.js").UWSClientOpts} UWSClientOpts
 */

/**
 * uWebSockets.js configuration options for UWSProxy's constructor.
 * @see {UWSProxy.createUWSConfig}
 * @typedef UWSProxyUWSConfigOpts
 * @property {boolean|null} [ssl=null]     Default: `null` - If true, inform the Proxy that trafic is encrypted (it
 *                                         matters to set proxy Headers and create (if not provided)
 *                                         an SSLApp instead of an App)
 * @property {int}           [port=443]    Default: `443` - Public port uWebSocket server is listening to
 * @property {boolean}       [quiet=false] Default: `false` - Disable configuration warning printing
 * @property {UWSAppOptions} [config=Object] Default: `{}` - See uWebSockets.js AppOptions
 */

/**
 * UWSProxy actual configuration for uWebSockets.js
 * @typedef UWSProxyUWSConfig
 * @property {boolean}               ssl
 * @property {int}                   port
 * @property {UWSProxyUWSConfigOpts} config
 * @property {UWSTemplatedApp}       server
 */

/**
 * Callbacks dictionary used by UWSProxy for you to hook / change the proxy behavior in some
 * circumstances.
 * @typedef UWSProxyCallbacks
 * @property {UWSProxyErrorHandler|null} [error=null] Default: `null` - Called when a proxy request fails for whatever reason.
 */

/**
 * UWSProxy configuration options.
 * @typedef UWSProxyOpts
 * @property {Object} [backpressure=Object]
 * @property {int}    [backpressure.maxStackedBuffers=4096] Default: `4096` - Once the buffer is filled, the connection
 *                                                          will be aborted
 * @property {Object<string, string|string[]>} [headers=Object] Additional headers always appended to
 *                                                          the proxy request (not on the client's
 *                                                          response)
 * @property {Object<string, string>} [routes] Routes we want the proxy request handlers to listen
 *                                             to
 * @property {int}    [timeout=300000] Default: `300000` - Timeout in MS before an attempt to reach the proxied server
 *                                        will abort.
 * @property {UWSProxyCallbacks} [on=Object] Collection of optional callbacks
 */

/**
 * Used by UWSProxy to send a proper Error response to the client if possible. This object can
 * be returned by a UWSProxyCallbacks.error callback to change the UWSProxy default error response.
 * @see {UWSProxyCallbacks.error}
 * @typedef UWSProxyErrorResponse
 * @property {string} status
 * @property {Object<string,string|string[]>} headers
 * @property {UWSRecognizedString} body
 */

/**
 * @typedef {import('./utils/uwsHelpers').UWSDecodedRequest} UWSDecodedRequest
 */

/**
 * @callback UWSProxyErrorHandler
 * Called on request error. Can be used as a hook to change the Proxy error response if an
 * UWSProxyErrorResponse is returned. **Supports promises**. This **can** return a promise or not.
 *
 * If no (or falsy) result is returned, the Proxy will respond with default error handling. This behavior
 * allows for custom logging or whatever.
 *
 * You may want to take those NodeJS error codes into consideration:
 *
 * - **ECONNRESET**
 * - **ECONNABORT**
 * - **ECONNREFUSED**
 * - **ETIMEDOUT**
 *
 * See [NodeJS Error Codes](https://nodejs.org/api/errors.html#errors_common_system_errors) for more information.
 *
 * The proxy may give you some custom codes:
 *
 * - **E_RECIPIENT_ABORTED**: the recipient server aborted either with partial or no response at all. It
 *   may or may not have received the request.
 *   In this scenario, the original error is passed to the Error constructor as cause,
 *   and the original error code (if any) can be found in error.original_error.
 * - **E_PIPELINE_ABORTED**: the pipeline was aborted before the request was sent to the recipient server,
 *   or before the response was received from the recipient server.
 * - **E_PIPELINE_OVERFLOW**: the pipeline aborted because the number of requests it received exceeded.
 *   increase maxPipelinedRequestsByConnection or maxConnectionsByHost to avoid this error
 * - **E_INVALID_CONTENT_LENGTH**: The response parser received a Content-Length header with an invalid value.
 * - **E_INVALID_CHUNK_LENGTH**: The response parser received a chunk length with an invalid value.
 * @param {Error} error
 * @param {UWSDecodedRequest} decodedRequest
 * @return {UWSProxyErrorResponse|void|Promise<UWSProxyErrorResponse|void>}
 */

// endregion

/**
 * A proxy based on uWebSockets.js. Allow for compatibility between uWebSockets.js and any HTTP server.
 *
 * It is useful in restricted server environment like clouds where you can't set up a proxy, or if you
 * don't want to use a proxy and need uWebSockets.js to work on the same port as any other http server
 * you're already using like express, nestjs, fastify, etc.
 *
 * **Example with express:**
 *
 * ```js
 * const http = require('http');
 * const express = require('express');
 * const uWebSockets = require('uWebSockets.js');
 *
 * const {
 * 	UWSProxy,
 * 	createUWSConfig,
 * 	createHTTPConfig
 * } = require('uws-reverse-proxy');
 *
 * const port = process.env.PORT || 80;
 *
 * const proxy = new UWSProxy(
 * 	createUWSConfig(
 * 		uWebSockets,
 * 		{ port }
 * 	)
 * );
 *
 * const expressApp = express();
 * expressApp.listen(
 * 	proxy.http.port,
 * 	proxy.http.host,
 * 	() => console.log(`HTTP Server listening at ${proxy.http.protocol}://${proxy.http.host}:${proxy.http.port}`)
 * );
 *
 * proxy.uws.server.ws({
 * 	upgrade : () => {
 * 		//...
 *     },
 * 	//...
 * });
 *
 * proxy.uws.server.listen('0.0.0.0', port, listening => {
 * 	if(listening){
 * 		console.log(`uWebSockets.js listening on port 0.0.0.0:${port}`);
 * 	}else{
 * 		console.error(`Unable to listen on port 0.0.0.0:${port}!`);
 * 	}
 * });
 * ```
 *
 * @see More examples in the [examples repository](https://github.com/jordan-breton/uws-reverse-proxy-examples)
 *
 */
class UWSProxy {

	// region Static methods

	/**
	 * Create a valid httpConfiguration. The purpose of this method is to emit warnings or throw errors
	 * if the configuration doesn't seem valid.
	 *
	 * It's also meant to be a helper for further updates, where new configuration options will be added.
	 *
	 * using this method to configure UWSProxy is strongly recommended.
	 *
	 * @param {UWSProxyHTTPConfigOpts}   [config=Object]  Configuration object
	 * @return {UWSProxyHTTPConfig}
	 */
	static createHTTPConfig(config = {}){
		const {
			port = 35974,
			host = '127.0.0.1',
			protocol = 'http',
			quiet = false,
			client = {}
		} = config || {};

		if(
			!quiet
			&& protocol === 'https'
			&& ['localhost', '127.0.0.1'].includes(host.toLowerCase().trim())
		){
			console.warn(
				'[WARN] UWSProxy: you configured the proxy to forward to a local HTTPS server.'
				+ ' You should consider using an HTTP server, as TLS have an impact on performances.'
			);
		}

		return {
			config,
			host,
			port,
			protocol,
			client: client || {}
		};
	}

	/**
	 * Creates a valid uWebSockets.js configuration. The purpose of this method is to emit warnings
	 * or throw errors if the configuration doesn't seem valid.
	 *
	 * It's also meant to be a helper for further updates, where new configuration options will be added.
	 *
	 * using this method to configure UWSProxy is strongly recommended.
	 *
	 * @param {UWS|UWSTemplatedApp} uWebSocket
	 * @param {UWSProxyUWSConfigOpts} opts
	 * @return {UWSProxyUWSConfig}
	 */
	static createUWSConfig(uWebSocket, opts = {}){
		if(!uWebSocket){
			throw new Error(
				"First argument required! Must be either the uWebSockets.js package itself"
				+ " (require('uWebSockets.js')) or an instance of"
				+ " uWebSockets.js:App / uWebSockets.js:SSLApp"
			);
		}

		const {
			App,
			SSLApp
		} = uWebSocket;

		let {
			ssl = null,
			port = 443,
			quiet = false,
			config = {}
		} = opts || {};

		let uwsServer;

		if(!App && !SSLApp){
			// If App and SSLApp are undefined, we try to determine if the first argument is a
			// constructed App or SSLApp. Since the uWebSockets.js package do not expose those
			// classes, we have to guess using a hacky way... It's not reliable because it may be
			// changed by the maintainer later, but it's all we have.
			if(!uWebSocket.constructor?.name?.startsWith('uWS.')){
				throw new Error(
					"The first argument doesn't seems to be a uWebSockets.js app"
					+ " nor the uWebSockets.js package itself."
				);
			}else{
				if(ssl === null){
					ssl = uWebSocket.constructor.name === 'uWS.SSLApp';
				}

				uwsServer = uWebSocket;
			}
		}else{
			if(ssl === null){
				ssl = UWS_SSL_KEYS.some(key => key in config);
			}

			if(ssl) uwsServer = SSLApp(config);
			else uwsServer = App(config);
		}

		if(!port || !Number.isInteger(port) || port < 2 || port > 49151){
			throw new Error('opts.ports must be a valid integer and a valid port number!');
		}else if(!opts.port && !quiet && port === 443 && !ssl){
			console.warn(
				"[WARN] UWSProxy: No port was specified in opts."
				+ " Default port used is 443."
			)
		}

		return {
			config,
			server: uwsServer,
			ssl,
			port
		}
	}

	// endregion
	// region Instance properties

	/**
	 * @type {UWSProxyUWSConfig}
	 */
	_uwsConfig;

	/**
	 * @type {UWSProxyHTTPConfig}
	 */
	_httpConfig;

	/**
	 * @type {UWSProxyOpts}
	 */
	_opts;

	/**
	 * @type {Client}
	 * @private
	 */
	_httpClient;

	// endregion

	/**
	 * @param {UWSProxyUWSConfig}  uwsConfig  uWebSockets.js configuration. You should create it
	 *                                        with UWSProxy.createUWSConfig
	 * @param {UWSProxyHTTPConfig} httpConfig HTTP configuration of the target HTTP server. You should create it
	 *                                        with UWSProxy.createHTTPConfig.
	 * @param {UWSProxyOpts}       opts Proxy configuration options.
	 */
	constructor(
		uwsConfig,
		httpConfig,
		opts = {}
	) {
		if(!uwsConfig) throw new Error('No uWebSockets.js configuration provided!');
		if(!httpConfig) httpConfig = UWSProxy.createHTTPConfig();

		const {
			routes = null,
			headers = {},
			timeout = 300000,
			on : {
				error = null
			} = {},
			backpressure: {
				maxStackedBuffers = 4096
			} = {}
		} = opts || {};

		this._uwsConfig = uwsConfig;
		this._httpConfig = httpConfig;
		this._opts = {
			backpressure: {
				maxStackedBuffers: typeof maxStackedBuffers === 'number' ? maxStackedBuffers : 4096
			},
			headers: headers || {},
			routes: routes || {
				any: '/*'
			},
			timeout: typeof timeout === 'number' ? timeout : 300000,
			on: {
				error
			}
		};

		this._httpClient = new HTTPClient(httpConfig.client);
	}

	// region Getters

	/**
	 * A shallow copy of the uWebSocket configuration.
	 *
	 * Note that the `config` property is the raw object passed as a parameter to
	 * UWSProxy.createUWSConfig.
	 * @return {UWSProxyUWSConfig}
	 */
	get uws(){
		const {
			config,
			server,
			ssl,
			port,
		} = this._uwsConfig;

		return {
			config,
			server,
			ssl,
			port
		};
	}

	/**
	 * A shallow copy of the http configuration
	 *
	 * Note that the `config` property is the raw object passed as a parameter to
	 * UWSProxy.createHTTPConfig.
	 * @return {UWSProxyHTTPConfig}
	 */
	get http(){
		const {
			config,
			host,
			port,
			protocol
		} = this._httpConfig;

		return {
			config,
			host,
			port,
			protocol
		};
	}

	// endregion

	/**
	 * Attach routes listeners to uWebSocket to start proxying.
	 *
	 * @important This action can't be undone. uWebSockets.js do not allow listeners removal.
	 */
	start(){
		const { routes } = this._opts;
		const { server: uwsServer } = this._uwsConfig;

		Object.keys(routes).forEach(method => {
			uwsServer[method](routes[method], this._handleRequest.bind(this));
		});
	}

	/**
	 * Handle a request received by uWebSockets.js and forward it to the http server.
	 * @param {UWSResponse} uwsResponse
	 * @param {UWSRequest} uwsRequest
	 */
	_handleRequest(uwsResponse, uwsRequest){
		const request = decodeRequest(uwsResponse, uwsRequest)

		const {
			host: privateHost,
			port: privatePort,
			protocol: privateProtocol
		} = this._httpConfig;

		const { headers: optsHeaders } = this._opts;

		uwsResponse.onAborted(() => {

			// We just destroy the body stream if any. We can't abort the request because
			// it's pipelined. So we will just ignore the response when we'll get it.
			uwsResponse.aborted = true;
		});

		this._httpClient.request({
			protocol: privateProtocol,
			host: privateHost,
			port: privatePort,
			path: request.url + '?' + request.query,
			method: request.method,
			headers: Object.assign(
				{},
				request.headers,
				optsHeaders
			),
			response: uwsResponse
		}, err => {
			if(err){
				this._tryToRespondToError(err, uwsResponse, request);
			}
		});
	}

	// region Error handling

	/**
	 * Construct a valid error response based on the provided error. The error response
	 * will be sent (if possible) to the client. To change any of the default response,
	 * use a UWSProxyErrorHandler callback
	 * @param {Error} error The error we want to build a response upon
	 * @return {UWSProxyErrorResponse}
	 */
	_buildErrorResponse(error){
		const response = {
			headers: {},
			body: undefined,
			status: undefined
		};

		switch(error.code){
			case 'ECONNRESET':
			case 'ECONNABORTED':
			case 'ECONNREFUSED':
			case 'E_PIPELINE_OVERFLOW':
				response.status = "503 Service Unavailable";
				response.body = `Unable to forward the request to the server (${error.code}).`;
				break;

			case 'ETIMEDOUT':
				response.status = "504 Gateway Timeout";
				response.body = `No response received from the server in ${this._opts.timeout}ms: request aborted (${error.code}).`;
				break;

			case 'E_PIPELINE_ABORTED':
				response.status = "502 Bad Gateway";
				response.body = `The request have been aborted by the proxy (${error.code}).`;
				break;

			case 'E_RECIPIENT_ABORTED':
				response.status = "502 Bad Gateway";
				response.body = `The recipient server aborted the proxy request (${error.code}).`;
				break;

			case 'E_INVALID_CONTENT_LENGTH':
			case 'E_INVALID_CHUNK_LENGTH':
				response.status = "502 Bad Gateway";
				response.body = `The proxy received a malformed or incomplete response from the server (${error.code}).`;
				break;

			default:

				// In every other case the response is invalid for a reason or another.
				response.status = "502 Bad Gateway";
				response.body = `The proxy encountered a non-handled error (${error.code}).`
		}

		return response;
	}

	/**
	 * Make an attempt to send an UWSProxyErrorResponse to the client. This may produce no result
	 * if the UWSResponse have been aborted/closed already. This method will just fail silently if it
	 * happens.
	 *
	 * @param {UWSResponse} uwsResponse The response we want to write into
	 * @param {UWSProxyErrorResponse} errorResponse The error to send.
	 */
	_tryToSendErrorResponse(uwsResponse, errorResponse){
		const {
			headers,
			body,
			status
		} = errorResponse;

		try{
			writeHeaders(uwsResponse, Object.assign(
				{},
				headers,
				{ status }
			));
		}catch(err){
			// We can ignore it, headers may have been sent already
		}

		try{
			uwsResponse.cork(() => {
				uwsResponse.end(body);
			});
		}catch(err){
			// We can ignore it, the uwsResponse has probably been closed already if we go there.
		}
	}

	/**
	 * Try to use the provided UWSProxyErrorHandler if it was specified in UWSProxyOpts at UWSProxy
	 * creation.
	 * @param {UWSResponse} uwsResponse The response we want to send an error into.
	 * @param {Error} error The raw error that have been detected.
	 * @param {UWSDecodedRequest} request Informations about the current request.
	 * @return {boolean} False if no handler is defined, true otherwise
	 */
	_tryToUseErrorHandlerResponse(uwsResponse, error, request){
		const {
			on: {
				error: errorHandler
			} = {}
		} = this._opts;

		if(!errorHandler) return false;

		const res = errorHandler(error, request);

		if(isPromise(res)){
			res.then((errorResponse) => {
				this._tryToSendErrorResponse(
					uwsResponse,
					errorResponse || this._buildErrorResponse(error)
				);
			}).catch(err => {

				// This should not happen, that's why we force a print to the console. If you want
				// remove this print, you must ensure your errorHandler is not throwing any error.
				console.error('UWSProxy: error thrown in error handler: ', err);
				this._tryToSendErrorResponse(uwsResponse, this._buildErrorResponse(error));
			});
		} else {
			this._tryToSendErrorResponse(uwsResponse, res || this._buildErrorResponse(error));
		}

		return true;
	}

	/**
	 * Will try to respond to an error, either by using a UWSProxyErrorHandler (if any) or by sending
	 * a default error response.
	 * @param {Error} error The error that have been detected.
	 * @param {UWSResponse} uwsResponse The response we want to write into.
	 * @param {UWSDecodedRequest} request Informations about the current request.
	 */
	_tryToRespondToError(error, uwsResponse, request){
		if(!this._tryToUseErrorHandlerResponse(uwsResponse, error, request)){
			this._tryToSendErrorResponse(uwsResponse, this._buildErrorResponse(error));
		}
	}

	// endregion
}

module.exports = UWSProxy;